File: php-bot/src/Services/SymbolService.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Database;

final class SymbolService
{
    /** @param string[] $seedSymbols */
    public static function ensureDefaults(array $seedSymbols): void
    {
        $pdo = Database::connection();
        $existing = array_column($pdo->query('SELECT symbol FROM symbols')->fetchAll(), 'symbol');
        $existingSet = array_flip($existing);
        foreach ($seedSymbols as $sym) {
            if (!isset($existingSet[$sym])) {
                $stmt = $pdo->prepare('INSERT INTO symbols (symbol, enabled, added_at) VALUES (:s, 1, :now)');
                $stmt->execute(['s' => $sym, 'now' => Database::now()]);
            }
        }
    }

    /** @return string[] */
    public static function listEnabled(): array
    {
        $rows = Database::connection()->query('SELECT symbol FROM symbols WHERE enabled = 1 ORDER BY symbol')->fetchAll();
        return array_column($rows, 'symbol');
    }

    /** @return array<int, array<string, mixed>> */
    public static function listAll(): array
    {
        return Database::connection()->query('SELECT * FROM symbols ORDER BY symbol')->fetchAll();
    }

    public static function addSymbol(string $symbol): void
    {
        $symbol = strtoupper(trim($symbol));
        $pdo = Database::connection();
        $stmt = $pdo->prepare('SELECT id FROM symbols WHERE symbol = :s');
        $stmt->execute(['s' => $symbol]);
        if ($row = $stmt->fetch()) {
            $upd = $pdo->prepare('UPDATE symbols SET enabled = 1 WHERE id = :id');
            $upd->execute(['id' => $row['id']]);
        } else {
            $ins = $pdo->prepare('INSERT INTO symbols (symbol, enabled, added_at) VALUES (:s, 1, :now)');
            $ins->execute(['s' => $symbol, 'now' => Database::now()]);
        }
    }

    public static function removeSymbol(string $symbol): bool
    {
        $stmt = Database::connection()->prepare('DELETE FROM symbols WHERE symbol = :s');
        $stmt->execute(['s' => $symbol]);
        return $stmt->rowCount() > 0;
    }

    public static function setEnabled(string $symbol, bool $enabled): bool
    {
        $stmt = Database::connection()->prepare('UPDATE symbols SET enabled = :e WHERE symbol = :s');
        $stmt->execute(['e' => $enabled ? 1 : 0, 's' => $symbol]);
        return $stmt->rowCount() > 0;
    }
}

File: app/services/symbol_service.py
from __future__ import annotations

from sqlalchemy import select

from app.core.database import get_session
from app.core.models import Symbol


async def ensure_defaults(seed_symbols: list[str]) -> None:
    async with get_session() as session:
        result = await session.execute(select(Symbol))
        existing = {s.symbol for s in result.scalars().all()}
        for sym in seed_symbols:
            if sym not in existing:
                session.add(Symbol(symbol=sym, enabled=True))


async def list_enabled() -> list[str]:
    async with get_session() as session:
        result = await session.execute(select(Symbol).where(Symbol.enabled == True).order_by(Symbol.symbol))  # noqa: E712
        return [s.symbol for s in result.scalars().all()]


async def list_all() -> list[Symbol]:
    async with get_session() as session:
        result = await session.execute(select(Symbol).order_by(Symbol.symbol))
        return list(result.scalars().all())


async def add_symbol(symbol: str) -> Symbol:
    symbol = symbol.upper().strip()
    async with get_session() as session:
        result = await session.execute(select(Symbol).where(Symbol.symbol == symbol))
        row = result.scalar_one_or_none()
        if row is None:
            row = Symbol(symbol=symbol, enabled=True)
            session.add(row)
        else:
            row.enabled = True
        await session.flush()
        await session.refresh(row)
        return row


async def remove_symbol(symbol: str) -> bool:
    async with get_session() as session:
        result = await session.execute(select(Symbol).where(Symbol.symbol == symbol))
        row = result.scalar_one_or_none()
        if row is None:
            return False
        await session.delete(row)
        return True


async def set_enabled(symbol: str, enabled: bool) -> bool:
    async with get_session() as session:
        result = await session.execute(select(Symbol).where(Symbol.symbol == symbol))
        row = result.scalar_one_or_none()
        if row is None:
            return False
        row.enabled = enabled
        return True
